pkg/registry/rbac/reconciliation/reconcile_clusterrolebindings.go

```diff
@@ -0,0 +1,260 @@
+		// If created since we started this reconcile, re-run
+		if errors.IsAlreadyExists(err) {
+			return o.run(attempts + 1)
+		}
+		if err != nil {
+			return nil, err
+		}
+		result.RoleBinding = created
+
+	case ReconcileUpdate:
+		updated, err := o.Client.Update(result.RoleBinding)
+		// If deleted since we started this reconcile, re-run
+		if errors.IsNotFound(err) {
+			return o.run(attempts + 1)
+		}
+		if err != nil {
+			return nil, err
+		}
+		result.RoleBinding = updated
+
+	case ReconcileNone:
+		// no-op
+
+	default:
+		return nil, fmt.Errorf("invalid operation: %v", result.Operation)
+	}
+
+	return result, nil
+}
+
+// computeReconciledRoleBinding returns the rolebinding that must be created and/or updated to make the
+// existing rolebinding's subjects, roleref, labels, and annotations match the expected rolebinding
+func computeReconciledRoleBinding(existing, expected *rbac.ClusterRoleBinding, removeExtraSubjects bool) (*ReconcileClusterRoleBindingResult, error) {
+	result := &ReconcileClusterRoleBindingResult{Operation: ReconcileNone}
+
+	result.Protected = (existing.Annotations[rbac.AutoUpdateAnnotationKey] == "false")
+
+	// Reset the binding completely if the roleRef is different
+	if !equivalentRoleRefs(expected.RoleRef, existing.RoleRef) {
+		result.RoleBinding = expected
+		result.Operation = ReconcileRecreate
+		return result, nil
+	}
+
+	// Start with a copy of the existing object
+	changedObj, err := api.Scheme.DeepCopy(existing)
+	if err != nil {
+		return nil, err
+	}
+	result.RoleBinding = changedObj.(*rbac.ClusterRoleBinding)
+
+	// Merge expected annotations and labels
+	result.RoleBinding.Annotations = merge(expected.Annotations, result.RoleBinding.Annotations)
+	if !reflect.DeepEqual(result.RoleBinding.Annotations, existing.Annotations) {
+		result.Operation = ReconcileUpdate
+	}
+	result.RoleBinding.Labels = merge(expected.Labels, result.RoleBinding.Labels)
+	if !reflect.DeepEqual(result.RoleBinding.Labels, existing.Labels) {
+		result.Operation = ReconcileUpdate
+	}
+
+	// Compute extra and missing subjects
+	result.MissingSubjects, result.ExtraSubjects = diffSubjectLists(expected.Subjects, existing.Subjects)
+
+	switch {
+	case !removeExtraSubjects && len(result.MissingSubjects) > 0:
+		// add missing subjects in the union case
+		result.RoleBinding.Subjects = append(result.RoleBinding.Subjects, result.MissingSubjects...)
+		result.Operation = ReconcileUpdate
+
+	case removeExtraSubjects && (len(result.MissingSubjects) > 0 || len(result.ExtraSubjects) > 0):
+		// stomp to expected subjects in the non-union case
+		result.RoleBinding.Subjects = expected.Subjects
+		result.Operation = ReconcileUpdate
+	}
+
+	return result, nil
+}
+
+func equivalentRoleRefs(a, b rbac.RoleRef) bool {
+	if a.APIGroup == "" || b.APIGroup == "" {
+		// If only one specified an APIGroup, don't compare that field
+		a.APIGroup = ""
+		b.APIGroup = ""
+	}
+	return a == b
+}
+
+func equivalentSubjects(a, b rbac.Subject) bool {
```

Comment:
this will simplify after https://github.com/kubernetes/kubernetes/pull/41184 merges